Minio/Helper/RequestUtil.cs

```diff
@@ -0,0 +1,161 @@
+            string endpointURL = string.Format("{0}://{1}", scheme, endpoint);
+            Uri uri = null;
+            try
+            {
+                uri = new Uri(endpointURL);
+            }
+            catch (UriFormatException e)
+            {
+                throw new InvalidEndpointException(e.Message);
+            }
+            return uri;
+        }
+
+        /// <summary>
+        /// Validates URI to check if it is well formed. Otherwise cry foul.
+        /// </summary>
+        public static void ValidateEndpoint(Uri uri,string Endpoint)
+        {
+            if (string.IsNullOrEmpty(uri.OriginalString))
+            {
+                throw new InvalidEndpointException("Endpoint cannot be empty.");
+            }
+            string host = uri.Host;
+
+            if (!isValidEndpoint(uri.Host))
+            {
+                throw new InvalidEndpointException(Endpoint, "Invalid endpoint.");
+            }
+            if (!uri.AbsolutePath.Equals("/", StringComparison.CurrentCultureIgnoreCase))
+            {
+                throw new InvalidEndpointException(Endpoint, "No path allowed in endpoint.");
+            }
+
+            if (!string.IsNullOrEmpty(uri.Query))
+            {
+                throw new InvalidEndpointException(Endpoint, "No query parameter allowed in endpoint.");
+            }
+            if ((!uri.Scheme.ToLowerInvariant().Equals("https")) && (!uri.Scheme.ToLowerInvariant().Equals("http")))
+            {
+                throw new InvalidEndpointException(Endpoint, "Invalid scheme detected in endpoint.");
+            }
+            string amzHost = uri.Host;
+            if ((utils.CaseInsensitiveContains(amzHost,".amazonaws.com"))
+                 && !s3utils.IsAmazonEndPoint(uri.Host))
+            {
+                throw new InvalidEndpointException(amzHost, "For Amazon S3, host should be \'s3.amazonaws.com\' in endpoint.");
+            }
+        }
+       
+        /// <summary>
+        /// Validate Url endpoint 
+        /// </summary>
+        /// <param name="endpoint"></param>
+        /// <returns>true/false</returns>
+        public static bool isValidEndpoint(string endpoint)
+        {
+            // endpoint may be a hostname
+            // refer https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names
+            // why checks are as shown below.
+            if (endpoint.Length < 1 || endpoint.Length > 253)
+            {
+                return false;
+            }
+
+            foreach (var label in endpoint.Split('.'))
+            {
+                if (label.Length < 1 || label.Length > 63)
+                {
+                    return false;
+                }
+
+                Regex validLabel = new Regex("^[a-zA-Z0-9][a-zA-Z0-9-]*");
+                Regex validEndpoint = new Regex(".*[a-zA-Z0-9]$");
+
+                if (!(validLabel.IsMatch(label) && validEndpoint.IsMatch(endpoint)))
```

Comment:
`endpoint` is not changing in the enclosing for-loop here. So this match with validEndpoint appears to be redundantly done. If it should be `validEndpoint.IsMatch(label)` then the regexps above can be merged into a single one.